### PR TITLE
[New Component] Service graph processor

### PR DESCRIPTION
**Pull request by mapno**
**Description:** First draft of the service graph processor.

The service graphs processor is a traces processor that builds a map representing the interrelationships between various services in a system.
The processor will analyse trace data and generate metrics describing the relationship between the services.
These metrics can be used by data visualization apps (e.g. Grafana) to draw a service graph.

Service graphs are useful for a number of use-cases:

* Infer the topology of a distributed system. As distributed systems grow, they become more complex. Service graphs can help you understand the structure of the system.
* Provide a high level overview of the health of your system. Service graphs show error rates, latencies, among other relevant data.
* Provide an historic view of a system’s topology. Distributed systems change very frequently,
and service graphs offer a way of seeing how these systems have evolved over time.

This component is based on [Grafana Tempo's service graph processor](https://github.com/grafana/tempo/tree/main/modules/generator/processor/servicegraphs).

Follow-up from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10425

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9232

**Testing:** Unit tests are added across all relevant files (mainly `processor_test.go`).

### CODE REVIEW

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:118]: Similar to a previous comment: I think the pipeline builder logs this already.

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:157]: md?

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:217]: What's 2 here? Is there a constant somewhere that can be used here?

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:277]: Perhaps it would be worth recording the edge data, like the trace ID, edge type and service name?

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:282]: Same here?

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:293]: If you don't need them to be serially updated, you can consider using an atomic integer instead of a lock + ints.

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:494]: Here it is! The readme says this would be seconds, but it looks like it is milliseconds. In any case, I think the config type can either be changed to be a slice of integers or the conversion should be removed altogether. Right now, it's a slice of durations, so users can in theory specify "1d" and this will be converted to float here.

File: processor/servicegraphprocessor/processor.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor.go:470]: Isn't there a cache TTL already specified somewhere? Should that be used instead?

File: processor/servicegraphprocessor/processor_test.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor_test.go:127]: Semantically speaking, shouldn't this test shut down the processor? It might also be good to leave a comment stating that the assertions are part of the verifyMetrics function, called in the "prepare" phase of the unit test.

File: processor/servicegraphprocessor/testdata/service-graph-config.yaml
Comment (jpkrohling) [inline processor/servicegraphprocessor/testdata/service-graph-config.yaml:9]: Should there be a metrics exporter here in the config?

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/internal/store/edge.go:31]: Are you using TraceID for any purposes? Perhaps use pcommon.TraceID instead of converting the []bytes to string?

File: processor/servicegraphprocessor/README.md
Comment (jpkrohling) [inline processor/servicegraphprocessor/README.md:100]: Having now gone through the PR, I see that the readme is quite outdated. Could you give it a second pass? For instance, the processor needs a reference to a metrics exporter, which is not mentioned here. Instead, the readme says that metrics are generated in Prometheus format.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md:67]: Added in 60cfaf7119db2023323328c0cef25e08c6e9afdf, nice suggestion.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: You're right, it was very updated. Fixed in 60cfaf7119db2023323328c0cef25e08c6e9afdf, thanks.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: Correct, fixed in 60cfaf7119db2023323328c0cef25e08c6e9afdf.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: Updated the comment in 60cfaf7119db2023323328c0cef25e08c6e9afdf.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: Removed, nice catch.

File: processor/servicegraphprocessor/config.go
Comment (mapno) [reply processor/servicegraphprocessor/config.go:28]: Updated 60cfaf7119db2023323328c0cef25e08c6e9afdf :)

File: processor/servicegraphprocessor/factory_test.go
Comment (mapno) [reply processor/servicegraphprocessor/factory_test.go]: Updated in 60cfaf7119db2023323328c0cef25e08c6e9afdf.

File: processor/servicegraphprocessor/go.mod
Comment (mapno) [reply processor/servicegraphprocessor/go.mod]: Updated, thanks

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/edge.go]: This was left from a previous implementation, removed!

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go]: Actually, this method is only used for testing, so no need to run with a mutex at all.

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go:106]: Ups, I completely missed implementing expiration. Fixed, thanks.

File: processor/servicegraphprocessor/internal/store/store_test.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store_test.go]: Updated, thanks

File: processor/servicegraphprocessor/processor_test.go
Comment (mapno) [reply processor/servicegraphprocessor/processor_test.go:130]: Agree, with the suggestion. Updated.

File: processor/servicegraphprocessor/testdata/service-graph-config.yaml
Comment (mapno) [reply processor/servicegraphprocessor/testdata/service-graph-config.yaml:9]: In a valid collector config, yes. In this case it's only validating that the processor's config is correct.

File: processor/servicegraphprocessor/factory_test.go
Comment (mapno) [reply processor/servicegraphprocessor/factory_test.go]: You're right, fixed it!

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/edge.go]: Right, it also saves converting to uint64 all the time. Changed in bdecf6c5e4e8ef4462fbac15f58fee0bcdd1717f

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go:32]: In turned out that `len()` didn't need a mutex after all.

**Review by mapno — COMMENTED** [review]
Thanks for the very thorough review @jpkrohling. Updated the branch addressing the comments. There were a couple of important things that I had missed 😅

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: Fixed!

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md:67]: Good point, added.

File: processor/servicegraphprocessor/config.go
Comment (mapno) [reply processor/servicegraphprocessor/config.go:28]: Nice catch, fixed.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: Bad copy pasting 😅 Fixed

File: processor/servicegraphprocessor/factory_test.go
Comment (mapno) [reply processor/servicegraphprocessor/factory_test.go]: Changed.

File: processor/servicegraphprocessor/go.mod
Comment (mapno) [reply processor/servicegraphprocessor/go.mod]: Updated 👍

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/edge.go]: It also saves converting to uint64 all the time. Nice.

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go]: `len()` is only called for testing and doesn't need the mutex at all.

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go:106]: Ups, I missed implementing the store expiration. Done!

File: processor/servicegraphprocessor/internal/store/store_test.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store_test.go]: Changed

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go]: Changed to use variable

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go]: Good point, added
/resolve

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go]: Good point, added

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go:294]: I reviewed the implementation again, they need to be updated all at the same time, b/c `keyToMetric` can be modified in between.

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/edge.go]: Changed to `pcommon.TraceID`.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: You're right. I've updated it with all the suggestions.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: You were right with the comments regarding this. The config admits `time.Duration`. Updated the comment.

File: processor/servicegraphprocessor/README.md
Comment (mapno) [reply processor/servicegraphprocessor/README.md]: They come from resource and span attributes. I updated the comment to better explain it.

File: processor/servicegraphprocessor/factory_test.go
Comment (mapno) [reply processor/servicegraphprocessor/factory_test.go]: This is now clarified 👍 Good catch

File: processor/servicegraphprocessor/internal/store/store_test.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store_test.go:145]: Yes. It's only ensuring that there are no concurrency problems between the two.

> If that's the case, would this also work by making a more precise incision with WaitGroups?

I'm not sure if there is a good way to do it with `WaitGroup`. Both functions take the store's lock, blocking mid-call would stall the execution of the test.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go]: Removed!

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go]: Removed!

File: processor/servicegraphprocessor/testdata/service-graph-config.yaml
Comment (mapno) [reply processor/servicegraphprocessor/testdata/service-graph-config.yaml:9]: Not really. This yaml is for a test checking that the processor's config is correct, not the entire pipeline.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go:484]: These are two different configs. The TTL you're mentioning is tracking how long edges are kept in the store before they're recorded even if incomplete. This cache cleanup is removing stale metrics that haven't received an update in 15 minutes.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go:508]: This should be fixed. The config admits a slice of `time.Duration`. Converting it to milliseconds makes it easier to work with metrics.

**Comment by jpkrohling** [comment]
Looks like there's a check failure: 

```
FAIL: "processor/servicegraphprocessor" not included in CODEOWNERS
```

I'll review the PR anyway, but please include this component as part of the code owners. Feel free to include yourself and me in there.

File: processor/servicegraphprocessor/internal/store/store.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/internal/store/store.go:110]: This looks incomplete, but looking closely, it's fine. Perhaps add a comment, stating that it will iterate until there are no more items to be evicted?

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Looks like there's one more CI failure: 

```
Validating versions.yaml
multimod verify
Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml
verifyAllModulesInSet failed: Module github.com/open-telemetry/opentelemetry-collector-contrib/processor/servicegraphprocessor (defined in /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/processor/servicegraphprocessor/go.mod) is not listed in any module set.
make: *** [Makefile:374: multimod-verify] Error 1
```

File: processor/servicegraphprocessor/internal/store/store.go
Comment (mapno) [reply processor/servicegraphprocessor/internal/store/store.go:111]: Good point, added in c182bebcf7306f152b153623b0fdc97b3648e2b3

**Comment by mapno** [comment]
Thanks for the second review. CI should be green now ✅

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Wonderful, thanks!

**Pull request merged**
